### PR TITLE
Add documentation to Hash

### PR DESCRIPTION
**Pull request by adlerhsieh**

### CODE REVIEW

File: src/hash.cr [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# A `Hash` represents a mapping of keys to values. 
+# Check [official docs](http://crystal-lang.org/docs/syntax_and_semantics/literals/hash.html) for the basics.
```
Comment (adlerhsieh) [inline src/hash.cr]: I'm not sure if putting a link in the introduction is a good idea or not. We don't duplicate the introduction by putting a link there, but it takes another click for users to check the documentation.

File: src/hash.cr [hunk 2]
```diff
@@ -430,6 +432,12 @@ class Hash(K, V)
     nil
   end
 
+  # Returns an `Array` populated with the results of each iteration in the given block.
+  #
+  # ```crystal
```
Comment (luislavena) [inline src/hash.cr]: According to [documenting code](http://crystal-lang.org/docs/conventions/documenting_code.html) there is no need for the language to be specified in the comment blocks as it assumes the code is already Crystal.

**Comment by adlerhsieh** [comment]
Updated. Thanks!

By the way, @kirbyfan64 it will be better if you comment the file in this pull request. You commented on my forked repo and this pull request look like:

screen shot 2015-10-14 at 10 12 54 am

It's was little hard to track the line number :P Thanks a lot!

File: src/hash.cr [hunk 3]
```diff
@@ -438,6 +446,15 @@ class Hash(K, V)
     array
   end
 
+  # Appends *other* to a `Hash`, but does not alter it.
```
Comment (jhass) [inline src/hash.cr]: I'm not sure append is the right word. And not just any `Hash`, but the receiver.

> Returns a new hash with the keys and values of this hash and \*other\* combined. A value in *other* takes precedence over the one in this hash.

Something like this? Similar below.

File: src/hash.cr [hunk 4]
```diff
@@ -643,6 +704,13 @@ class Hash(K, V)
     to_s(io)
   end
 
+  # Converts to a `String`, preserving all symbols including {}, :, and =>.
```
Comment (jhass) [inline src/hash.cr]: I think that's slightly misleading, since `{"foo": "bar"}.to_s # =>{"foo" => "bar"} "

File: src/hash.cr [hunk 5]
```diff
@@ -677,6 +746,11 @@ class Hash(K, V)
     end
   end
 
+  # Inverts keys and values.
```
Comment (jhass) [inline src/hash.cr]: Maybe mention that in case of duplicated values the last value's key becomes the new value?

File: src/hash.cr [hunk 6]
```diff
@@ -685,6 +759,16 @@ class Hash(K, V)
     hash
   end
 
+  # Returns *true* if all key-value pairs returns *true* in the given block.
```
Comment (jhass) [inline src/hash.cr]: > Yields all key-value pairs to the given block, returns `true` if the block returns a truthy value for all key-value pairs, else `false`.

?

File: src/hash.cr [hunk 7]
```diff
@@ -699,10 +793,31 @@ class Hash(K, V)
...
   end
 
+  # Returns *true* if a `Hash` has any key-value pair.
   def any?
     !empty?
   end
 
+  # Iterates over each key-value pair with the given block, 
+  # replacing *memo* with the returned value each time. 
+  # Returns *memo* after all iteration.
```
Comment (jhass) [inline src/hash.cr]: Mention the argument more explicitly, specifically that it provides the initial value?

> Returns the last value of \*memo\*.

?

File: src/hash.cr [hunk 8]
```diff
@@ -699,10 +793,31 @@ class Hash(K, V)
...
+  #
+  # ```
+  # hash = {
+  #   "foo":   "bar",
+  #   "hello": "world"
+  # }
+  #
+  # hash.inject(StringIO.new) do |r,k,v|
+  #   r << " #{k} #{v}"
+  # end
```
Comment (jhass) [inline src/hash.cr]: I don't think that's a good example, this one is better for `each_with_object`.

File: src/hash.cr [hunk 9]
```diff
@@ -699,10 +793,31 @@ class Hash(K, V)
...
+  # #=> " foo bar hello world"
+  #
+  # hash.inject(0) do |r,k,v|
+  #   r += v.size
```
Comment (jhass) [inline src/hash.cr]: That assignment is misleading since it's superfluous.

**Comment by jhass** [comment]
As usual I'd love to see some space after `,` in the examples.

**Comment by kirbyfan64** [comment]
@adlerhsieh Crud...I thought I did...I rarely ever use GitHub's line notes, so I probably goofed it up. Sorry!

**Comment by adlerhsieh** [comment]
Updated.

File: src/hash.cr [hunk 10]
```diff
@@ -452,6 +470,13 @@ class Hash(K, V)
     hash
   end
 
+  # Returns a new hash with the keys and values of this hash and *other* combined.
```
Comment (jhass) [inline src/hash.cr]: This is not true. It does not return a new hash, `merge` does. `merge!` modifies the receiver.

File: src/hash.cr [hunk 11]
```diff
@@ -556,23 +582,35 @@ class Hash(K, V)
...
     first = @first.not_nil!
     {first.key, first.value}
   end
 
+  # Returns the first key in the hash.
   def first_key
     @first.not_nil!.key
   end
 
+  # Returns the first key if it exists, or returns nil.
```
Comment (jhass) [inline src/hash.cr]: \`nil\`

File: src/hash.cr [hunk 12]
```diff
@@ -643,6 +704,13 @@ class Hash(K, V)
     to_s(io)
   end
 
+  # Converts to a `String`, with all symbols escaped, including {}, :, and =>. 
+  #
+  # ```
+  # h = {"foo": "bar"}
+  # h.to_s       #=> "{\"foo\": \"bar\"}"
```
Comment (jhass) [inline src/hash.cr]: This is still wrong, it does not return this.

File: src/hash.cr [hunk 13]
```diff
@@ -643,6 +704,13 @@ class Hash(K, V)
     to_s(io)
   end
 
+  # Converts to a `String`, with all symbols escaped, including {}, :, and =>.
```
Comment (jhass) [inline src/hash.cr]: I'm really not sure we should talk about symbols and escaping here. "Returns a string representation" should be enough IMO.

File: src/hash.cr [hunk 14]
```diff
@@ -1,3 +1,6 @@
+# A `Hash` represents a mapping of keys to values. 
+# 
+# Check [official docs](http://crystal-lang.org/docs/syntax_and_semantics/literals/hash.html) for the basics.
```
Comment (jhass) [inline src/hash.cr]: "See the ..."

File: src/hash.cr (hunk 13 above)
Comment (adlerhsieh) [inline src/hash.cr]: OK. I'll make it simple.

**Comment by asterite** [comment]
@jhass Now the formatter also formats code inside doc comments, so hopefully life will be easier starting from the next release :-)

**Comment by adlerhsieh** [comment]
Updated. Thanks for clarifying the misunderstanding.

File: src/hash.cr [hunk 15]
```diff
@@ -643,6 +705,13 @@ class Hash(K, V)
     to_s(io)
   end
 
+  # Converts to a `String`.
+  #
+  # ```
+  # h = {"foo": "bar"}
+  # h.to_s       #=> {"foo": "bar"}
```
Comment (jhass) [inline src/hash.cr]: This example is still not right.

**Comment by adlerhsieh** [comment]
Updated. I finally got the point. Thanks.

**Pull request merged**